### PR TITLE
Support TIMESTAMPNS

### PR DESCRIPTION
**Pull request by WiSaGaN**
This adds support of linux TIMESTAMPNS.
The code is mostly copied paste from https://github.com/nix-rust/nix/pull/663

### CODE REVIEW

**Review by asomers — CHANGES_REQUESTED** [review]
It looks pretty good so far.  But why did you decide to write the test without the usual test harness?  It's preferable to use the regular test harness if possible.  In fact, right now your test won't even be compiled, because you didn't add it to Cargo.toml.

**Comment by WiSaGaN** [comment]
> It looks pretty good so far. But why did you decide to write the test without the usual test harness? It's preferable to use the regular test harness if possible. In fact, right now your test won't even be compiled, because you didn't add it to Cargo.toml.

Thanks for the review! Is there a way to use the macro `cmsg_space` in unit tests? It contains `use unix::...`, which I can't make it work within a unit test.

File: CHANGELOG.md
Comment (asomers) [inline CHANGELOG.md:8]: Why only x86_64 and x86?  It looks like that symbol is defined for many other arches, too.

**Review by asomers — COMMENTED** [review]
I assume you mean `cmsg_space!`?  It's currently used by the functional tests in test/sys/test_socket.rs.  There aren't any unit tests that currently use it, but I don't see why there couldn't be.  What problem are you having?

**Comment by WiSaGaN** [comment]
> I assume you mean `cmsg_space!`? It's currently used by the functional tests in test/sys/test_socket.rs. There aren't any unit tests that currently use it, but I don't see why there couldn't be. What problem are you having?

It contains self-referential "use unix::*" statement, which can't be used in unit tests.

File: CHANGELOG.md
Comment (WiSaGaN) [reply CHANGELOG.md:8]: Even now arch like `Linux i686` is still failing. I am wondering why.

**Comment by asomers** [comment]
> > I assume you mean `cmsg_space!`? It's currently used by the functional tests in test/sys/test_socket.rs. There aren't any unit tests that currently use it, but I don't see why there couldn't be. What problem are you having?
> 
> It contains self-referential "use unix::*" statement. Apparently, when in "nix" crate, you can only reference it as "use crate::..." but not "use nix:...".

Well, that sounds like something you can fix.

**Comment by WiSaGaN** [comment]
> > > I assume you mean `cmsg_space!`? It's currently used by the functional tests in test/sys/test_socket.rs. There aren't any unit tests that currently use it, but I don't see why there couldn't be. What problem are you having?
> > 
> > 
> > It contains self-referential "use unix::*" statement. Apparently, when in "nix" crate, you can only reference it as "use crate::..." but not "use nix:...".
> 
> Well, that sounds like something you can fix.

I think it would be better to do that in a separate pull request. In the mean time, I can add a todo of moving the integration test to unit test after that is done.

File: Cargo.toml
Comment (asomers) [inline Cargo.toml:80]: Why put it in its own test program, as opposed to the main "test" executable?  Only tests that require their own process should get their own test programs.  It's fine for the test to be a functional test; I don't think there's any reason to prefer a unit test instead.  But it shouldn't have its own test program.

File: CHANGELOG.md
Comment (WiSaGaN) [reply CHANGELOG.md]: It seems that even i686 linux does not have that symbol. Neither did mips, etc...

I have restricted to x86_64, and now the tests pass.

File: src/sys/socket/mod.rs
Comment (asomers) [inline src/sys/socket/mod.rs:562]: You should not restrict the target_arch to just x86_64.  libc defines SCM_TIMESTAMPNS for many architectures.

**Review by asomers — CHANGES_REQUESTED** [review]

File: Cargo.toml
Comment (WiSaGaN) [reply Cargo.toml]: This is the CI result when the conditional compilation was turned off: https://cirrus-ci.com/build/5978723427024896

I think there are cases where the symbols are added, but it was not supported: https://cirrus-ci.com/task/5721153533640704?command=main#L510

File: src/sys/socket/mod.rs
Comment (WiSaGaN) [reply src/sys/socket/mod.rs:562]: This is the CI result when the conditional compilation was turned off: https://cirrus-ci.com/build/5978723427024896

I think there are cases where the symbols are added, but it was not supported: https://cirrus-ci.com/task/5721153533640704?command=main#L510

Comment (asomers) [reply src/sys/socket/mod.rs:562]: It may be that Linux supports that socket option, but QEMU does not.  Looking at the Linux source code, I think it should be supported everywhere.

Comment (WiSaGaN) [reply src/sys/socket/mod.rs:562]: I can add more arch if you have concrete ones in mind. Otherwise I think it makes sense to merge this as MVP since x86_64 covers major use cases, and follow up with other archs.

Comment (asomers) [reply src/sys/socket/mod.rs:562]: You should add every arch that is supported by libc.  Don't be deterred by the fact that you can't run it in CI; some architectures don't have any CI coverage at all.  You have a working test case and there is no reason to believe that the feature would work differently on other architectures.

Comment (WiSaGaN) [reply src/sys/socket/mod.rs:562]: Adding those would break CI, which prevents us from getting breaking warnings for future changes for those archs. Are you sure that is what we want?

Comment (asomers) [reply src/sys/socket/mod.rs:562]: I'm not asking you to break CI.  I'm asking you to get the feature working on all platforms.  But if QEMU doesn't support it, then disable the test (not the feature) on those platforms.  There are already plenty of examples of tests that are disabled on QEMU platforms.

**Comment by WiSaGaN** [comment]
`libc` recently updated its socketopt consts: https://github.com/rust-lang/libc/commit/682eba61169005a6d904fc53fc749788f7e3cf9b

We will wait for the release of this to have a more updated lists of arch here.

**Comment by WiSaGaN** [comment]
After upgrade libc to 0.2.93, now the support has been added to all linux platform. Tests however are restricted to x86-64, non musl due to suspected QEMU issue.

**Comment by WiSaGaN** [comment]
Please feel free to squash all the commits into one if it is ready to merge.

File: test/sys/test_socket.rs
Comment (asomers) [inline test/sys/test_socket.rs:1541]: This can't be right.  CI doesn't use QEMU for x86-64-linux-musl.  What error do you get on that platform?

Comment (WiSaGaN) [reply test/sys/test_socket.rs]: Fixed the leftover from before the libc upgrade.

**Comment by asomers** [comment]
Ok, this looks good now.  But you'll have to squash yourself.  Nix uses the bors merge bot, which is incompatible with Github's squash-and-merge button.

**Comment by WiSaGaN** [comment]
Sqaushed. Now it is ready to merge.

**Review by asomers — APPROVED** [review]
bors r+

**Pull request merged**
